### ISSUE TITLE
how to create a custom validation and register in root

### ISSUE THREAD

**Issue by GlauberF** [ngx-formly/ngx-formly#2059]
how to create a custom validations file and register in FormlyModule.forRoot

I already have the personalized message file and I already register
![Captura de Tela_selecionar área_20200129174006](https://user-images.githubusercontent.com/11314585/73395524-7a390c80-42be-11ea-9e69-812691a8b1f6.png)
![Captura de Tela_selecionar área_20200129174034](https://user-images.githubusercontent.com/11314585/73395510-70afa480-42be-11ea-8d3e-680bce9914bf.png)

I want to create a custom validation like this https://formly.dev/examples/validation/matching-two-fields

however, I want to be able to inform which field should correspond, making it dynamic, which I can use for various scenarios, one of which is password and confirm password

**Comment by aitboudad** [comment]
you can get the current field instance from the second argument of the message validation function

**Comment by GlauberF** [comment]
@aitboudad yes, but I couldn't register inside FormlyModule.forRoot in `validators: []`

**Comment by aitboudad** [comment]
so I think  the request is not clear, to register the `fieldMatch` you have to pass validation callback to `validators` and the message through the `validationMessages`

```ts
    FormlyModule.forRoot({
      validators: [
        { name: 'fieldMatch', validation: fieldMatchValidator },
      ],
      validationMessages: [
        { name: 'fieldMatch', message: fieldMatchValidatorMessage },
      ],
    }),
```

**Comment by GlauberF** [comment]
ok I registered now.
![Captura de Tela_selecionar área_20200129183252](https://user-images.githubusercontent.com/11314585/73399346-d7848c00-42c5-11ea-93bd-7e6d4e7824ad.png)

but after validations, how to return inside
![Captura de Tela_selecionar área_20200129183347](https://user-images.githubusercontent.com/11314585/73399392-e8cd9880-42c5-11ea-8b5b-8b826710ba3d.png)

**Comment by aitboudad** [comment]
if you want to return a message from the validation itself, I think it's possible by returning an object with `message` property:

```ts
return { message: 'Your message' }
```

**Comment by GlauberF** [comment]
ok, but i already have the message registered within the providers, why should i return the message again?
I did not understand!

![Captura de Tela_selecionar área_20200129185011](https://user-images.githubusercontent.com/11314585/73400644-51b61000-42c8-11ea-8561-b634a394ec20.png)
![Captura de Tela_selecionar área_20200129185049](https://user-images.githubusercontent.com/11314585/73400649-54186a00-42c8-11ea-8e98-5b9d367a6255.png)

**Comment by aitboudad** [comment]
the following question is not clear:

> but after validations, how to return inside

**Comment by GlauberF** [comment]
ok, I have a function `VimboFieldMatchValidator` it should return true or false only.
```javascript
export function VimboFieldMatchValidator(control, field): FieldValidatorFn {
    const value = control.value;

    return value[field.key] === value[field.templateOptions.analyzeFieldMatch]
        // avoid displaying the message error when values are empty
        || (!field.templateOptions.analyzeFieldMatch || !value[field.key] || !value[field.templateOptions.analyzeFieldMatch]);
}
```
![Captura de Tela_20200129192241](https://user-images.githubusercontent.com/11314585/73402753-bc694a80-42cc-11ea-9340-cc953cece717.png)
```javascript
FormlyModule.forRoot({
validators: [
                {
                    name: 'fieldMatch',
                    validation: VimboFieldMatchValidator
                }
            ]
})
```

and the message to be shown I created inside `registerFormlyValidationMessages`, which is registered in the module in providers.
```javascript
export function registerFormlyValidationMessages(): any {
    return {
        validationMessages: [
            {
                name: 'fieldMatch',
                message: formlyValidationMessage('fieldMatch')
            }
        ]
    };
}
```
![Captura de Tela_selecionar área_20200129185049](https://user-images.githubusercontent.com/11314585/73402135-56c88e80-42cb-11ea-8a8d-c5c6e489e889.png)

put it in `VimboFieldMatchValidator`, I can't return true or false, it asks to return something like.
```javascript
{
    [key: string]: any;
};
```
then the question, what should I return in `VimboFieldMatchValidator`

**Comment by aitboudad** [comment]
if `true` return `null` else return an object `{}`.

```ts
const isValid = value[field.key] === value[field.templateOptions.analyzeFieldMatch]
        // avoid displaying the message error when values are empty
        || (!field.templateOptions.analyzeFieldMatch || !value[field.key] || !value[field.templateOptions.analyzeFieldMatch]);

return isValid ? null : { fieldMatch: true };

**Comment by GlauberF** [comment]
even doing the way you went, he accuses the return type of error.
![Captura de Tela_20200130085158](https://user-images.githubusercontent.com/11314585/73443431-654b9000-4335-11ea-8dd1-277f78a922b0.png)

I believe that if I remove the type above (FieldValidatorFn) it will work, but it is not the correct one to be done.

**Comment by aitboudad** [comment]
The function expects `ValidationErrors` and not `FieldValdiatorFn`.

Please check our example which includes an example of custom validation through NgModule declaration https://formly.dev/examples/validation/custom-validation + next time a `stackblitz` example is required to avoid asking a lot of questions :)

**Comment by GlauberF** [comment]
next time, i will provide a stackblitz!

I managed to register, but I still have a little problem, the configuration of the fields are like this.

```javascript
{
                key: 'emp_senha_cer',
                type: 'input',
                templateOptions: {
                    type: 'password',
                    label: 'Senha (PIN)',
                    required: true,
                    appearance: 'outline',
                    enableWorkCount: true
                }
            },
            {
                key: 'emp_senha_cer_confirm',
                type: 'input',
                templateOptions: {
                    type: 'password',
                    label: 'Confirmar Senha (PIN)',
                    required: true,
                    appearance: 'outline',
                    enableWorkCount: true,
                    analyzeFieldMatch: 'emp_senha_cer'
                },
                validators: {
                    validation: ['fieldMatch'],
                }
            }
```

where in the field (emp_senha_cer_confirm) I specify the field that should be used as a basis of comparison `analyzeFieldMatch: 'emp_senha_cer'`

However, within the `VimboFieldMatchValidator` validation function, I do not receive the value of the field to be compared to` emp_senha_cer`

```javascript
import { FormControl, ValidationErrors } from '@angular/forms';
import { FormlyFieldConfig } from '@ngx-formly/core/lib/components/formly.field.config';

export function VimboFieldMatchValidator(control: FormControl, field: FormlyFieldConfig): ValidationErrors {
    console.log('control', control);
    console.log('field', field);
    // return null if you have not passed the analyzeFieldMatch setting,
    // it means that this field does not need this validation
    if (!field.templateOptions.analyzeFieldMatch) {
        return null;
    }
    // if it has no value, returns null
    if (!control.value) {
        return null;
    }
    // Get Value
    const value = control.value;
    console.log('>> value <<', value);
    // Check if is valid
    const isValid =
        value[field.key] === value[field.templateOptions.analyzeFieldMatch]
        // avoid displaying the message error when values are empty
        || !value[field.key] || !value[field.templateOptions.analyzeFieldMatch];

    console.log('>> isValid <<', isValid);
    // return validator
    return isValid ? null : { fieldMatch: true }
}
```

in other words, I need to get the value of `emp_senha_cer` to be able to compare the two` emp_senha_cer_confirm === emp_senha_cer `within the` VimboFieldMatchValidator` validation function

**Comment by GlauberF** [comment]
done, thanks!

```javascript
import { FormControl, ValidationErrors } from '@angular/forms';
import { FormlyFieldConfig } from '@ngx-formly/core/lib/components/formly.field.config';

export function VimboFieldMatchValidator(
    control: FormControl,
    fieldConfig: FormlyFieldConfig
): ValidationErrors {

    if (!control.parent || !control) {
        return null;
    }

    // return null if you have not passed the analyzeFieldMatch setting,
    // it means that this field does not need this validation
    if (!fieldConfig.templateOptions.analyzeFieldMatch) {
        return null;
    }

    // Compare fields
    const field = control.parent.get(fieldConfig.key);
    const fieldConfirm = control.parent.get(fieldConfig.templateOptions.analyzeFieldMatch);

    // Avoid displaying the message error when values are empty
    if (!field || !fieldConfirm) {
        return null;
    }

    // if (fieldConfirm.value === '') {
    //     return null;
    // }

    // Equal
    if (field.value === fieldConfirm.value) {
        return null;
    }

    // return validator
    return { fieldMatch: true };
}
```

**Comment by aitboudad** [comment]
This issue has been fixed and released as part of [v5.7](https://github.com/ngx-formly/ngx-formly/releases/tag/v5.7.0) release.

Please let us know, in case you are still encountering a similar issue/problem.
Thank you!
